OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,779 @@
+// persistent, so that we can try to read the timestamps on a present.
+// Note that we only store the state into `QueueSubmission`, if at that time we have a value for the
+// end_info. So once it is submitted, the end_info will always be set.
+// Beside the information about the begin/end, it also stores the text, color and the depth of the
+// marker. If a debug marker begin was discarded because of its depth, cut_off is set to true.
+// This allows end markers on a different submission to also throw the end marker away.
+// Example: Max Depth = 1
+// Submission 1: Begin("Foo"), Begin("Bar) -- For "Bar" we set cut-off to true.
+// Submission 2: End("Bar"), End("Foo") -- We now know, that the first end needs to be thrown away.
+struct MarkerState {
+  std::optional<SubmittedMarker> begin_info;
+  std::optional<SubmittedMarker> end_info;
+  std::string label_name;
+  Color color;
+  size_t depth;
+  bool cut_off;
+};
+
+// A single submission (VkQueueSubmit) can contain multiple `SubmitInfo`s. We keep this structure.
+struct SubmitInfo {
+  std::vector<SubmittedCommandBuffer> command_buffers;
+};
+
+// Wraps up all the data that needs to be persistent upon a submission (VkQueueSubmit).
+// `completed_markers` are all the debug markers, that got completed (via "End") within this
+// submission. Their "Begin" might still be in a different submission.
+struct QueueSubmission {
+  SubmissionMetaInformation meta_information;
+  std::vector<SubmitInfo> submit_infos;
+  std::vector<MarkerState> completed_markers;
+  uint32_t num_begin_markers = 0;
+};
+
+}  // namespace internal
+
+/*
+ * This class is responsible to track command buffer and debug marker timings.
+ * To do so, it keeps track of command-buffer allocations, destructions, begins, ends as well as
+ * submissions.
+ * I we are capturing on `VkBeginCommandBuffer` and `VkEndCommandBuffer` it will insert write
+ * timestamp commands (`VkCmdWriteTimestamp`). The same is done for debug marker begins and ends.
+ * All that data will be gathered together at a queue submission (`VkQueueSubmit`).
+ *
+ * Upon every `VkQueuePresentKHR` it will check if the timestamps of a certain submission are
+ * already available, and if so, it will send the results over to the `VulkanLayerProducer`.
+ *
+ * See also `DispatchTable` (for vulkan dispatch), `TimerQueryPool` (to manage the timestamp slots),
+ * and `DeviceManager` (to retrieve device properties).
+ *
+ * Thread-Safety: This class is internally synchronized (using read/write locks), and can be
+ * safely accessed from different threads. This is needed, as in Vulkan submits and command buffer
+ * modifications can happen from multiple threads.
+ */
+template <class DispatchTable, class DeviceManager, class TimerQueryPool>
+class SubmissionTracker : public VulkanLayerProducer::CaptureStatusListener {
+ public:
+  explicit SubmissionTracker(uint32_t max_local_marker_depth_per_command_buffer,
```

Comment:
Probably just a preference, but I would put the `max_local_marker_depth_per_command_buffer` last. It just seems a detail that should come after `dispatch_table`, `timer_query_pool`, `device_manager`.

Reply:
Done.